lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt

```diff
@@ -0,0 +1,351 @@
+    }
+
+    private fun updateMultipleEndpointsStatus(endpointIds: List<String>, status: EndpointStatus) {
+        _endpointStatusFlow.update { currentMap ->
+            currentMap.toMutableMap().apply {
+                endpointIds.forEach { endpointId ->
+                    this[endpointId]?.status = status
+                    this[endpointId]?.isOutgoing = true
+                }
+            }
+        }
+    }
+
+    private fun sendMmcpPingPacket(endpointId: String) {
+        val mmcpPing = MmcpPing(Random.nextInt())
+        val pingBytes = mmcpPing.toBytes()
+
+        val header = VirtualPacketHeader(
+            fromAddr = InetAddress.getByName(virtualIpAddress).address.ip4AddressToInt(),
+            toAddr = InetAddress.getByName(broadcastAddress).address.ip4AddressToInt(),
+            fromPort = 0,
+            toPort = 0,
+            lastHopAddr = InetAddress.getByName(virtualIpAddress).address.ip4AddressToInt(),
+            payloadSize = pingBytes.size,
+            hopCount = 0,
+            maxHops = 1
+        )
+
+        val totalSize = VirtualPacketHeader.HEADER_SIZE + pingBytes.size
+        val data = ByteArray(totalSize)
+        header.toBytes(data, 0)
+        System.arraycopy(pingBytes, 0, data, VirtualPacketHeader.HEADER_SIZE, pingBytes.size)
+
+        val virtualPacket = VirtualPacket.fromData(data, 0)
+        val payload = Payload.fromBytes(virtualPacket.data)
+
+        connectionsClient.sendPayload(endpointId, payload)
+            .addOnSuccessListener { log("Sent MMCP Ping to $endpointId") }
+            .addOnFailureListener { e -> log("Failed to send MMCP Ping to $endpointId", e) }
+    }
+
+    private fun sendMmcpPongPacket(endpointId: String, replyToMessageId: Int) {
+        val mmcpPong = MmcpPong(Random.nextInt(), replyToMessageId)
+        val pongBytes = mmcpPong.toBytes()
+
+        val header = VirtualPacketHeader(
+            fromAddr = InetAddress.getByName(virtualIpAddress).address.ip4AddressToInt(),
+            toAddr = InetAddress.getByName(broadcastAddress).address.ip4AddressToInt(),
+            fromPort = 0,
+            toPort = 0,
+            lastHopAddr = InetAddress.getByName(virtualIpAddress).address.ip4AddressToInt(),
+            payloadSize = pongBytes.size,
+            hopCount = 0,
+            maxHops = 1
+        )
+
+        val totalSize = VirtualPacketHeader.HEADER_SIZE + pongBytes.size
+        val data = ByteArray(totalSize)
+        header.toBytes(data, 0)
+        System.arraycopy(pongBytes, 0, data, VirtualPacketHeader.HEADER_SIZE, pongBytes.size)
+
+        val virtualPacket = VirtualPacket.fromData(data, 0)
+        val payload = Payload.fromBytes(virtualPacket.data)
+
+        connectionsClient.sendPayload(endpointId, payload)
+            .addOnSuccessListener { log("Sent MMCP Pong to $endpointId") }
+            .addOnFailureListener { e -> log("Failed to send MMCP Pong to $endpointId", e) }
+    }
+
+    private fun Int.toByteArray(): ByteArray {
```

Comment:
Please do not duplicate functions we already have that are already there, tested, and consistented. Look at the existing code - e.g. VirtualNode and use the existing functions.